cirq-google/cirq_google/devices/grid_device_test.py

```diff
@@ -0,0 +1,228 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest.mock as mock
+
+import pytest
+
+import cirq
+import cirq_google
+from cirq_google.api import v2
+
+
+GRID_HEIGHT = 5
+
+
+def _create_device_spec_with_horizontal_couplings():
+    # Qubit layout:
+    #   x -- x
+    #   x -- x
+    #   x -- x
+    #   x -- x
+    #   x -- x
+
+    grid_qubits = [cirq.GridQubit(i, j) for i in range(GRID_HEIGHT) for j in range(2)]
+
+    spec = v2.device_pb2.DeviceSpecification()
+    spec.valid_qubits.extend([v2.qubit_to_proto_id(q) for q in grid_qubits])
+    grid_targets = spec.valid_targets.add()
+    grid_targets.name = '2_qubit_targets'
+    grid_targets.target_ordering = v2.device_pb2.TargetSet.SYMMETRIC
+    for row in range(int(GRID_HEIGHT / 2)):
+        new_target = grid_targets.targets.add()
+        new_target.ids.extend([v2.qubit_to_proto_id(cirq.GridQubit(row, j)) for j in range(2)])
+    for row in range(int(GRID_HEIGHT / 2), GRID_HEIGHT):
+        # Flip the qubit pair order for the second half of qubits
+        # to verify GridDevice properly handles pair symmetry.
+        new_target = grid_targets.targets.add()
+        new_target.ids.extend([v2.qubit_to_proto_id(cirq.GridQubit(row, 1 - j)) for j in range(2)])
+    gate = spec.valid_gates.add()
+    gate.syc.SetInParent()
+    gate.gate_duration_picos = 12000
+    gate.valid_targets.extend(['2_qubit_targets'])
+
+    return grid_qubits, spec
+
+
+def _create_device_spec_with_all_couplings():
+    # Qubit layout:
+    #   x -- x
+    #   |    |
+    #   x -- x
+    #   |    |
+    #   x -- x
+    #   |    |
+    #   x -- x
+    #   |    |
+    #   x -- x
+
+    grid_qubits, spec = _create_device_spec_with_horizontal_couplings()
+    for row in range(GRID_HEIGHT - 1):
+        for col in range(2):
+            new_target = spec.valid_targets[0].targets.add()
+            new_target.ids.extend(
+                [
+                    v2.qubit_to_proto_id(cirq.GridQubit(row, col)),
+                    v2.qubit_to_proto_id(cirq.GridQubit(row + 1, col)),
+                ]
+            )
+    return grid_qubits, spec
+
+
+def _create_device_spec_qubit_pair_self_loops() -> v2.device_pb2.DeviceSpecification:
```

Comment:
Optional:  consider adding a docstring for this and the below private functions for improved clarity.